src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs

```diff
@@ -0,0 +1,149 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Globalization;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using MediatR;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Health.Abstractions.Exceptions;
+using Microsoft.Health.CosmosDb.Features.Storage;
+using Microsoft.Health.Fhir.Core.Features.Context;
+using Microsoft.Health.Fhir.CosmosDb.Features.Metrics;
+
+namespace Microsoft.Health.Fhir.CosmosDb.Features.Storage
+{
+    public class CosmosResponseProcessor : ICosmosResponseProcessor
+    {
+        private readonly IFhirRequestContextAccessor _fhirRequestContextAccessor;
+        private readonly IMediator _mediator;
+        private readonly ILogger<CosmosResponseProcessor> _logger;
+
+        public CosmosResponseProcessor(IFhirRequestContextAccessor fhirRequestContextAccessor, IMediator mediator, ILogger<CosmosResponseProcessor> logger)
+        {
+            EnsureArg.IsNotNull(fhirRequestContextAccessor, nameof(fhirRequestContextAccessor));
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _fhirRequestContextAccessor = fhirRequestContextAccessor;
+            _mediator = mediator;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Adds request charge to the response headers and throws a <see cref="RequestRateExceededException"/>
+        /// if the status code is 429.
+        /// </summary>
+        /// <param name="ex">The exception</param>
+        public async Task ProcessException(Exception ex)
+        {
+            if (_fhirRequestContextAccessor.FhirRequestContext == null)
+            {
+                return;
+            }
+
+            EnsureArg.IsNotNull(ex, nameof(ex));
+
+            if (ex is DocumentClientException dce)
+            {
+                await ProcessResponse(null, dce.RequestCharge, null, dce.StatusCode);
+
```

Comment:
Not sure if this is a change you made or if it was already like that, but could you put the parameter name for the two nulls for clarity?